File: backend/src/adapters/tokenbucket.py
import typing as ty
from functools import cached_property

from src.adapters.cache import KeySpace, RedisBool, RedisCache, ScriptFunc

type TokenBucketScript = ScriptFunc[list[str], list[float | int], RedisBool]


class Throttler(ty.Protocol):
    async def acquire(self, cost: int = 1) -> bool:
        ...


class TokenBucket:
    "Refill tokens with refill rate evertime it gets called"

    def __init__(
        self,
        redis: RedisCache,
        bucket_script: TokenBucketScript,
        *,
        bucket_key: KeySpace,
        max_tokens: int,
        refill_rate_s: float,
    ):
        self._redis = redis
        self._bucketscript = bucket_script

        # These three are instance attributes
        self._bucket_key = bucket_key
        self._max_tokens = max_tokens
        self._refill_rate_s = refill_rate_s

    @cached_property
    def _refill_token_script(
        self,
    ) -> ScriptFunc[list[str], list[float | int], RedisBool]:
        # only refills if bucket is not full
        lua = """
        -- Refill token bucket algorithm
        -- Keys: [bucket_key]
        -- Args: [token_cost, max_tokens]

        local bucket_key = KEYS[1]
        local token_cost = tonumber(ARGV[1])
        local max_tokens = tonumber(ARGV[2])

        -- Increment the token count
        local current_tokens = tonumber(redis.call('HINCRBY', bucket_key, 'tokens', token_cost))

        -- Ensure the token count does not exceed the max limit
        if current_tokens > max_tokens then
            redis.call('HSET', bucket_key, 'tokens', max_tokens)
        end

        return true
        """
        return self._redis.load_script(lua)

    async def release(self, cost: int = 1):
        res = await self._refill_token_script(
            keys=[self._bucket_key.key], args=[cost, self._max_tokens]
        )
        return res == 1

    async def acquire(self, cost: int = 1) -> ty.Literal[0] | float:
        args = [self._max_tokens, self._refill_rate_s, cost]
        wait_time = await self._bucketscript(keys=[self._bucket_key.key], args=args)
        return wait_time

    async def reserve_token(self, token_cost: int = 1) -> None:
        raise NotImplementedError


class TokenBucketFactory:
    def __init__(
        self,
        redis: RedisCache,
        script: TokenBucketScript,
        keyspace: KeySpace | None = None,
    ):
        self.cache = redis
        self.script = script
        self.keyspace = keyspace

    def create_bucket(
        self, bucket_key: str, max_tokens: int, refill_rate_s: float
    ) -> TokenBucket:
        key_ = self.keyspace(bucket_key) if self.keyspace else KeySpace(bucket_key)
        return TokenBucket(
            self.cache,
            self.script,
            bucket_key=key_,
            max_tokens=max_tokens,
            refill_rate_s=refill_rate_s,
        )

File: backend/askgpt/script/tokenbucket.lua
-- Token bucket algorithm implementation
-- Keys: [bucket_key]
-- Args: [max_tokens, refill_rate_s, token_cost]

local bucket_key, max_tokens, refill_rate_s, token_cost = KEYS[1], tonumber(ARGV[1]), tonumber(ARGV[2]),
    tonumber(ARGV[3])

-- Retrieve the current state from Redis
local bucket = redis.call('HMGET', bucket_key, 'last_refill_time', 'tokens')
local last_refill_time, tokens = tonumber(bucket[1]), tonumber(bucket[2])

-- Initialize if not present
if not last_refill_time then
    last_refill_time, tokens = redis.call('TIME')[1], max_tokens
end

-- Calculate tokens based on the elapsed time
local current_time = redis.call('TIME')[1]
local elapsed = current_time - last_refill_time
local new_tokens = math.min(max_tokens, tokens + elapsed * refill_rate_s)

-- Check if there are enough tokens
if new_tokens >= token_cost then
    local token_left = new_tokens - token_cost
    redis.call('HMSET', bucket_key, 'last_refill_time', current_time, 'tokens', token_left)
    return 0
else
    local needed = token_cost - new_tokens
    local wait_time = needed / refill_rate_s
    return wait_time
end
